### PR TITLE
[sharktank] e2e gpt-oss

### PR DESCRIPTION
**Pull request by oyazdanb**

### CODE REVIEW

File: sharktank/sharktank/examples/export_paged_llm_v1.py [hunk 1]
```diff
@@ -132,7 +133,14 @@ def _(model: ServicePagedLlmModelV1, tokens, seq_lens, seq_block_ids, cs):
     def generate_batch_decode(bs: int):
         # torch.export.Dim would make min at least 2
         block_dim_min = 2
-        block_dim_max = ceildiv(hp.context_length, llama_config.block_seq_stride) - 1
+        if hp.sliding_window and hp.sliding_window > 0:
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/examples/export_paged_llm_v1.py]: 0 is the default value.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 2]
```diff
@@ -347,7 +372,6 @@ def get_custom_configs(p: dict[str, Any], name_prefix: str):
         res["use_base_frequency_scaling"] = True
         res["use_fused_qkv"] = True
         res["topk_then_softmax"] = True
-        res["use_decomposed_attention"] = True
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/configs/llm_configs.py:350]: I have this value lingering from a branch and it's not needed or used here.

File: sharktank/sharktank/layers/ffn_moe_block.py [hunk 3]
```diff
@@ -86,6 +86,26 @@ def one_hot_matmul(self, inputs, weights, experts):
         output = einsum_2args(oh, matmul, "bm,bmn->mn")
         return output
 
+    def _broadcast_bias(self, bias, experts):
```
[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/layers/ffn_moe_block.py:89]: Right, I'm asking that you split this out in a separate PR with tests and land it before this one.

**Review by Alex-Vasile — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/ops/attention_impls.py [hunk 4]
```diff
@@ -73,6 +73,9 @@ def create_mask_sliding_window(
 
 def create_mask(a, attn_weights, is_causal):
     if a is not None:
+        # Ensure mask has same dtype as attn_weights to avoid dtype promotion
+        if a.dtype != attn_weights.dtype:
+            a = a.to(attn_weights.dtype)
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/ops/attention_impls.py:78]: the mask `a` is just a metadata and not model weights. if its dtype does not match `attn_weight`, pytorch will silently promote the result and break the dtype consistency. this check ensures `a` always matches the weights' dtype.

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 5]
```diff
@@ -116,6 +116,13 @@ def __init__(
                     f"{attn_var}_quantizer",
                     theta.optional_tensor(f"{attn_name}.q_output"),
                 )
+        self.sink = torch.nn.Parameter(
+            torch.empty(self.head_count, dtype=config.attention_dtype),
+            requires_grad=False,
+        )
+        if "attn_sinks" in theta.keys:
+            with torch.no_grad():
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/paged_llama_attention_block.py:124]: Do we need `torch.no_grad()`? Is there a sharktank equivalent feature we can rely on? We want to use in built sharktank components as much as possible.

File: sharktank/sharktank/ops/attention_impls.py [hunk 6]
```diff
@@ -128,7 +132,7 @@ def scaled_dot_product_attention_decomposed(
         kv_size=kv_size,
         sliding_window=sliding_window,
         dtype=q.dtype,
-        device=attn_weights.device,
+        device=q.device,  # Use q.device instead of attn_weights.device for consistency
```
Comment (archana-ramalingam) [inline sharktank/sharktank/ops/attention_impls.py:135]: Can remove comment here.

File: sharktank/sharktank/layers/paged_llama_attention_block.py [hunk 7]
```diff
@@ -116,6 +116,13 @@ def __init__(
                     f"{attn_var}_quantizer",
                     theta.optional_tensor(f"{attn_name}.q_output"),
                 )
+        self.sink = torch.nn.Parameter(
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/paged_llama_attention_block.py]: for now, I switched to .register_buffer but in  the following pr, i'm gonna make a new layer for sink.

File: sharktank/sharktank/models/llama/testing.py [hunk 8]
```diff
@@ -8,7 +8,7 @@
 import functools
 import torch
 import re
-
+import math
```
Comment (Alex-Vasile) [inline sharktank/sharktank/models/llama/testing.py:11]: Is this needed?

File: sharktank/tests/evaluate/datasets/llama_405b_fp8_e4m3fn_iree.json
Comment (Alex-Vasile) [inline sharktank/tests/evaluate/datasets/llama_405b_fp8_e4m3fn_iree.json:23]: Why are these being updated as part of this PR?

File: scripts/setenv.sh
Comment (Alex-Vasile) [inline scripts/setenv.sh:116]: How come these two need to be removed?

File: sharktank/sharktank/ops/attention_impls.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alex-Vasile) [reply sharktank/sharktank/ops/attention_impls.py]: Is this an issue that came up only for gpt-oss? Or is it a problem with other models?

[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/ops/attention_impls.py]: i noticed this in gpt-oss.

[the remark this answers is not in the document]
Comment (rsuderman) [reply sharktank/sharktank/ops/attention_impls.py]: Type changes should be deliberate and not done to suppress errors. It is likely this will just cause unknown errors downstream. Please focus on incremental feature work instead of monolithic landing. Seeing a patch titled `e2e gpt-oss` still indicates its trying to do full model instead of landing the separate features it requires.,

File: sharktank/sharktank/layers/paged_attention.py [hunk 9]
```diff
@@ -904,6 +904,15 @@ def paged_attention(
                 page_ids=seq_block_ids,
             )
 
+        # Ensure all tensors are in the correct attention dtype after kv cache read
+        # Only apply to unquantized tensors - quantized tensors will be handled in attention()
+        if not isinstance(q, (QuantizedTensor, PlanarQuantizedTensor)):
```
Comment (Alex-Vasile) [inline sharktank/sharktank/layers/paged_attention.py:909]:
```suggestion
        if not isinstance(q, QuantizedTensor):
```
Don't need to include subclasses
